### PR TITLE
yaml validator

### LINKED ISSUE #629: Normalize yaml strings
We've had problems when using service names to name resources like albs and rds db names. Whenever we unmarshal a `.yml` file we should check that name strings use only a-z,A-Z and hypens. We may need a special case for application name, since currently it's used in the naming of s3 buckets. See also: https://github.com/Originate/exosphere/issues/628

### PR DESCRIPTION
**Pull request by hugoesthere**
resolves: https://github.com/Originate/exosphere/issues/629

### CODE REVIEW

File: src/types/app_config.go [hunk 1]
```diff
@@ -125,3 +126,18 @@ func (a AppConfig) forEachService(fn func(string, string, ServiceData)) {
 		fn("public", serviceName, data)
 	}
 }
+
+func (a AppConfig) validateAppConfig() error {
+	appNameRegex := regexp.MustCompile("^[a-z0-9]+(-[a-z0-9]+)*$")
+	if !appNameRegex.MatchString(a.Name) {
+		return errors.New("only lowercase alphanumeric character(s) separated by a single hyphen are allowed. Must match regex: /^[a-z0-9]+(-[a-z0-9]+)*$/")
```
Comment (charlierudolph) [inline src/types/app_config.go]: Can this error message include the fact that error occurs in `application.yaml`, that its the `name` field, and the offending value?

Maybe something like. `The "name" field "invalid-app" in "application.yml" is invalid. ...`

File: src/types/app_config.go [hunk 2]
```diff
@@ -125,3 +126,18 @@ func (a AppConfig) forEachService(fn func(string, string, ServiceData)) {
...
+	}
+	var err error
+	serviceRoleRegex := regexp.MustCompile("^[a-zA-Z0-9]+(-[a-zA-Z0-9]+)*$")
+	a.forEachService(func(serviceType, serviceRole string, data ServiceData) {
+		if !serviceRoleRegex.MatchString(serviceRole) {
+			err = errors.New("only alphanumeric character(s) separated by a single hyphen are allowed. Must match regex: /^[a-zA-Z0-9]+(-[a-zA-Z0-9]+)*$/")
```
Comment (charlierudolph) [inline src/types/app_config.go]: Can this error message include that it is in `application.yml` and that its under `services.{serviceType}` and the offending key?

Maybe something like `The "services.public" key "invalid-service!" in "application.yml" is invalid. ...`

File: src/types/app_config_test.go [hunk 3]
```diff
@@ -40,6 +40,16 @@ var _ = Describe("AppConfig", func() {
...
 	})
 
+	var _ = Describe("NewAppConfig", func() {
+		It("should throw and error if any fields are invalid", func() {
+			appDir := path.Join("..", "..", "example-apps", "invalid-app-config")
+			_, err := types.NewAppConfig(appDir)
+			Expect(err).To(HaveOccurred())
+			expectedErrorString := "only alphanumeric character(s) separated by a single hyphen are allowed. Must match regex: /^[a-zA-Z0-9]+(-[a-zA-Z0-9]+)*$/"
+			Expect(err.Error()).To(ContainSubstring(expectedErrorString))
+		})
```
Comment (charlierudolph) [inline src/types/app_config_test.go]: Can we please have separate tests for an invalid app name and an invalid service name?

**Comment by hugoesthere** [comment]
@charlierudolph comments addressed

**Review by charlierudolph — APPROVED** [review]

**Pull request merged**